Support for large numbers of args/kws with Python 3.10

njriasan opened the pull request:
<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
forum https://numba.discourse.group/.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities here, please click the arrow besides
   "Create Pull Request" and choose "Create Draft Pull Request".
   When it's ready for review, you can click the button "ready to review" near
   the end of the pull request
   (besides "This pull request is still a work in progress".)
   The maintainers will then be automatically notified to review it.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on main/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against main they should
   be resolved by merging main into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

Closes #7812. This seems to be working for the set of tests I have covered. I think the code quality probably needs to improve considerably. In addition any code to test for unintended side effects is greatly appreciated.

benwilliamgraham reviewed (dismissed):
Looks good!

ehsantn commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,432 @@ def get_assignment_source(self, destname):
+                    raise UnsupportedError(errmsg)
+                # Determine the kws
+                if keyword_def.value.items:
+                    # n_kws <= 15 case.
```
I think it would be helpful to add example IRs in comments for each case that is being matched here.

njriasan replied:
Done!

ehsantn commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,432 @@ def get_assignment_source(self, destname):
+                            vararg_loc
+                        )
+                    else:
+                        args = _call_function_ex_replace_args_large(
```
Add comment on the expected IR node for args_def.

njriasan replied:
I added a comment explaining my assumptions. Please let me know if it needs further clarification.

ehsantn commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,432 @@ def get_assignment_source(self, destname):
+        MAP_ADD # Append the (key, value) pair to the map
+        # End for each argument
+
+    In the IR generated, the inital build map is empty and a series
```
```suggestion
    In the IR generated, the initial build map is empty and a series
```

ehsantn reviewed (commented):
Thanks for the effort @njriasan ! I took a quick look and the code makes sense to me.

njriasan commented:
@ehsantn and @benwilliamgraham thanks for doing an initial review.

@stuartarchibald could you mark this PR as ready for review? Thanks.

sklam commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,488 @@ def get_assignment_source(self, destname):
+            ):
+                not_found_getattr = False
+            else:
+                search_start += 1
```
Why is this needed and what would trigger this line?

njriasan replied:
This is necessary if some intermediate operation is needed to create the argument. For example, if one of the args was `arg5 = bool(n)`, then search_start skips over the function call.

I can update the comment and the tests to include a function call.

sklam commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,488 @@ def get_assignment_source(self, destname):
+        new_body[search_end] = None
+        search_end -= 1
+    else:
+        target_name = vararg_stmt.target.name
```
Why is this needed and what would trigger this line?

njriasan replied:
This may never be needed, but I wasn't sure the code will always produce an assignment (although I can't construct a case where it hasn't happened).

Basically the existing IR looks like:

```python
...
$final_var = build_tuple(items=[Var(argn, test.py:6)])
$final_tuple = $prev_tuple + $final_var
$varargs_var = $final_tuple
```
but it would be completely valid if instead the IR generated was
```python
...
$final_var = build_tuple(items=[Var(argn, test.py:6)])
$varargs_var = $prev_tuple + $final_var
```
so I was trying to protect against this ever changing.

njriasan replied:
If you are confident this extra assignment can't occur I'm happy to remove it.

sklam replied:
The extra assignment is coming from https://github.com/numba/numba/blob/59fa2e335be68148b3bd72a29de3ff011430038d/numba/core/interpreter.py#L259-L260. 

Let's make this branch raise an error to indicate it is unreachable and add a comment using your description in https://github.com/numba/numba/pull/7866#discussion_r817233996.

sklam commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,488 @@ def get_assignment_source(self, destname):
+                # We cannot handle this format.
+                raise UnsupportedError(errmsg)
+            if arg_tuple_stmt.target.name == lhs_name:
+                target_name = rhs_name
```
Impossible case? cpython doesn't seem to put the tuple on the LHS.

njriasan replied:
I agree I also never saw this on the LHS. I was simply trying to be conservative against any slight changes in the generated code. Since this is a property of the bytecode should I just remove this case?

sklam replied:
Let's make this raise an error indicating it is unreachable/impossible.

sklam commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,488 @@ def get_assignment_source(self, destname):
+                ):
+                    keep_looking = False
+                else:
+                    search_end -= 1
```
Why is this needed and what would trigger this line?

njriasan replied:
Same reason as the KWS case. If there is a function call like `bool(n)`

njriasan replied:
If there is anything need to "create" the arg, for example a function call.

sklam reviewed (commented):
I did a quick check on coverage and found the following lines to be never covered. I only checked a random 20% of the test suite and all the new tests thus the coverage data may not be 100% accurate. If these lines are not really dead, they should have more comments to explain them.

njriasan commented:
> I did a quick check on coverage and found the following lines to be never covered. I only checked a random 20% of the test suite and all the new tests thus the coverage data may not be 100% accurate. If these lines are not really dead, they should have more comments to explain them.

Thanks @sklam. I'll update my tests to add checks to two of the regions and I'll expand the comments on those regions.

Two of the areas that are uncovered were my attempts to be conservative against minor changes in CPython or the IR generated because I wasn't sure if my tests are exhaustive. If you feel like they are unnecessary I'm happy to remove them.

njriasan commented:
> > I did a quick check on coverage and found the following lines to be never covered. I only checked a random 20% of the test suite and all the new tests thus the coverage data may not be 100% accurate. If these lines are not really dead, they should have more comments to explain them.
> 
> Thanks @sklam. I'll update my tests to add checks to two of the regions and I'll expand the comments on those regions.
> 
> Two of the areas that are uncovered were my attempts to be conservative against minor changes in CPython or the IR generated because I wasn't sure if my tests are exhaustive. If you feel like they are unnecessary I'm happy to remove them.

@sklam I pushed an update to the tests and updated my comments.

stuartarchibald commented:
I don't think at present the transform will be able to handle inline control flow at the call site. For example, the rather contrived:
```python
import numba
import dis

@numba.njit
def my_func1(a=None, b=None, c=None, d=None, e=None, f=None, g=None, h=None, i=None, j=None, k=None, l=None, m=None, n=None, o=None, p=None):
    return my_func2(
        a=a,
        b=b,
        c=c,
        d=d,
        e=e,
        f=f,
        g=g,
        h=h,
        i=i,
        j=j,
        k=k,
        l=l,
        m=m,
        n=n,
        o=o,
        p = False if a else True
    )


@numba.njit
def my_func2(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p):
    return

print(dis.dis(my_func1.py_func))
print(my_func1())
```
which gives a bytecode that has:
```
  6          96 LOAD_CONST              16 ('p')

 22          98 LOAD_FAST                0 (a)
            100 POP_JUMP_IF_FALSE       55 (to 110)
            102 LOAD_CONST              17 (False)
            104 MAP_ADD                  1

  6         106 CALL_FUNCTION_EX         1
            108 RETURN_VALUE

 22     >>  110 LOAD_CONST              18 (True)
            112 MAP_ADD                  1

  6         114 CALL_FUNCTION_EX         1
```
at the location where `p` is resolved and added to the map prior to the call.

stuartarchibald commented:
> @stuartarchibald Thank you for finding these cases! You are definitely correct and I need to work on a more complete solution.

No problem.

I think given the rarity of the occurrence of these patterns, it would acceptable to do the following:
* Update this patch to catch the cases where control flow occurs. Inform users that control flow on arguments within call sites is not supported and refer them to docs on how to adjust their code.
* Add tests for the above. 

The most common use cases of this somewhat rare pattern would then be covered and workarounds for the uncommon use cases of this rare patter would be shown to users. Essentially, this patch makes the rare patterns work much more often, so I think it's better that it is reviewed and merged with acknowledgement of the imperfection instead of trying to cover everything. This of course also doesn't preclude further fixes later. What do you think?

njriasan commented:
> > @stuartarchibald Thank you for finding these cases! You are definitely correct and I need to work on a more complete solution.
> 
> No problem.
> 
> I think given the rarity of the occurrence of these patterns, it would acceptable to do the following:
> 
> * Update this patch to catch the cases where control flow occurs. Inform users that control flow on arguments within call sites is not supported and refer them to docs on how to adjust their code.
> * Add tests for the above.
> 
> The most common use cases of this somewhat rare pattern would then be covered and workarounds for the uncommon use cases of this rare pattern would be shown to users. Essentially, this patch makes the rare patterns work much more often, so I think it's better that it is reviewed and merged with acknowledgement of the imperfection instead of trying to cover everything. This of course also doesn't preclude further fixes later. What do you think?

@stuartarchibald I think this approach is reasonable and I agree with it.

njriasan commented:
@sklam I made your requested changes.

@stuartarchibald I made your changes as well and added tests. Proper error handling requires some extra code, so there may need to be to be a more detailed review. I also added a test for objmode which fails when the definitions are not properly updated.

sklam commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,647 @@ def get_assignment_source(self, destname):
+                vararg_loc = i - 1
+                args_def = None
+                not_found = True
+                while vararg_loc >= 0 and not_found:
+                    args_def = blk.body[vararg_loc]
+                    if (
+                        isinstance(args_def, ir.Assign)
+                        and args_def.target.name == vararg.name
+                    ):
+                        not_found = False
+                    else:
+                        vararg_loc -= 1
```
(Leaving thought trail in case we need it in the future. This probably doesn't present to be a problem)

With bytecode from:

```python
args = (*(0, 1, 2),)   # the starunpack triggers LIST_TO_TUPLE
y = my_func2(*args)
```
```
 16           0 BUILD_LIST               0
              2 LOAD_CONST               1 ((0, 1, 2))
              4 LIST_EXTEND              1
              6 LIST_TO_TUPLE
              8 STORE_FAST              31 (args)

 17          10 LOAD_GLOBAL              0 (my_func2)
             12 LOAD_FAST               31 (args)
             14 CALL_FUNCTION_EX         0
             16 STORE_FAST              32 (y)
```
the code here scan back from the call to `my_func2` to find the definition of `args`, which is suppose to be a `LIST_TO_TUPLE` opcode, but is replaced by `peep_hole_list_to_tuple`; otherwise, the later check will reject this call.

sklam replied:
I have also tried to trick this part of the code with the walrus operator but Numba is already miscompiling this on mainline:

```python
my_func2(args:=(*(0, 1, 2),), *args)
```

sklam commented on numba/core/interpreter.py:
```diff
@@ -79,7 +79,38 @@ def get_assignment_source(self, destname):
+        func_ir._definitions[lhs].clear()
+    else:
+        raise UnsupportedError(
+            "Inconsistency found the in definitions trying to remove varkwargs"
```
Is the error message still accurate? This function seem to be applied generally not just on varkwargs.

njriasan replied:
How is this message? Also thank you for the ping, I missed the original review.

sklam commented on numba/core/interpreter.py:
```diff
@@ -104,7 +104,9 @@ def _remove_assignment_definition(old_body, idx, func_ir):
     else:
         raise UnsupportedError(
-            "Inconsistency found the in definitions trying to remove varkwargs"
+            "Inconsistency found in the definitions while electuting"
```
```suggestion
            "Inconsistency found in the definitions while executing"
```

stuartarchibald commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+    # All changes are local to the a single block
+    # so it can be traversed in any order.
+    errmsg = """
+CALL_FUNCTION_EX with **kwargs not supported.
+If you are not using **kwargs this may indicate that
+you have a large number of kwargs and are using inlined control
+flow. You can resolve this issue by moving the control flow out of
+the function call. For example, if you have
+
+    f(a=1 if flag else 0, ...)
+
+Replace that with
+
+    a_val = 1 if flag else 0
+    f(a=a_val, ...)
```
Perhaps `textwrap.dedent` would let this be indented at a more expected level?

njriasan replied:
Agreed. Thank you!

stuartarchibald commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+                # varkwarg should be defined second so we start there.
+                varkwarg_loc = start_search
+                keyword_def = None
+                not_found = True
```
```suggestion
                found = False
```
(and fix up)?

stuartarchibald commented on numba/tests/test_interpreter.py:
```diff
@@ -0,0 +1,641 @@
+            'You can resolve this issue by moving the control flow out',
+            str(raises.exception)
+        )
+        # Uncomment sections when inlined control flow is
+        # actually supported.
+        # py_func = inline_func
+        # cfunc = njit()(inline_func)
+        # a = py_func(False)
+        # b = cfunc(False)
+        # self.assertEqual(a, b)
```
Can probably just remove this, it's easy to add in if/when needed.

stuartarchibald commented on numba/tests/test_interpreter.py:
```diff
@@ -0,0 +1,641 @@
+            'You can resolve this issue by moving the control flow out',
+            str(raises.exception)
+        )
+        # Uncomment sections when inlined control flow is
+        # actually supported.
+        # py_func = inline_func
+        # cfunc = njit()(inline_func)
+        # a = py_func(False)
+        # b = cfunc(False)
+        # self.assertEqual(a, b)
```
Remove? Same reasoning as above.

njriasan replied:
Done.

stuartarchibald commented on numba/tests/test_interpreter.py:
```diff
@@ -0,0 +1,641 @@
+        # self.assertEqual(a, b)
+
+    @skip_unless_py10
+    def test_large_args_uninlined_controlflow(self):
```
```suggestion
    def test_large_args_noninlined_controlflow(self):
```
(suggest same change throughout).

njriasan replied:
Done.

stuartarchibald commented on numba/tests/test_interpreter.py:
```diff
@@ -0,0 +1,641 @@
+            'You can resolve this issue by moving the control flow out',
+            str(raises.exception)
+        )
+        # Uncomment sections when inlined control flow is
+        # actually supported.
+        # py_func = inline_func
+        # cfunc = njit()(inline_func)
+        # a = py_func(False)
+        # b = cfunc(False)
+        # self.assertEqual(a, b)
```
Remove?

njriasan replied:
Done.

stuartarchibald reviewed (commented):
I've re-reviewed this patch following the changes to catch problems associated with control flow at call sites. It now seems to be handling the cases outlined in the original ticket as expected and failing in a controlled manner else where. Given @sklam has run coverage on this patch I've not checked this further. There's a few minor things to look at in line and once resolved I think this should be ready to merge. Thanks again for your efforts on this @njriasan.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+    """
+    Extracts the kws args passed as varkwarg
+    for CALL_FUNCTION_EX. This pass is taken when
+    n_kws <= 15 and the bytecode looks like:
```
This is an error in the docstring. Thank you for catching this.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+        # End for each argument
+
+    In the IR generated, the initial build map is empty and a series
+    of setitems are applied afterwards. THE IR looks like:
```
You are correct. I will update the description.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+        key_val = const_stmt.value.value
+        search_start += 1
+        # Now we need to search for a getattr with setitem
+        not_found_getattr = True
```
Agreed.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+            # original error message.
+            raise UnsupportedError(errmsg)
+        setitem_stmt = old_body[search_start + 1]
+        if not (
+            isinstance(setitem_stmt, ir.Assign)
+            and isinstance(setitem_stmt.value, ir.Expr)
+            and setitem_stmt.value.op == "call"
+            and (
+                setitem_stmt.value.func.name
+                == getattr_stmt.target.name
+            )
+            and len(setitem_stmt.value.args) == 2
+            and (
+                setitem_stmt.value.args[0].name
+                == key_var_name
+            )
+        ):
+            # We cannot handle this format so raise the
+            # original error message.
+            raise UnsupportedError(errmsg)
```
Actually for now I will just add the comment. If you would like me to change to an AssertionError I would be happy to, but I'm not certain this is actually unreachable.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+    together several 1 elem tuples to an initial empty tuple.
+    We traverse backwards in the IR, collecting args, until we
+    find the original empty tuple. For example, the IR might
+    look like:
+
+        $orig_tuple = build_tuple(items=[])
+        $first_var = build_tuple(items=[Var(arg0, test.py:6)])
+        $next_tuple = $orig_tuple + $first_var
+        ...
+        $final_var = build_tuple(items=[Var(argn, test.py:6)])
+        $final_tuple = $prev_tuple + $final_var
+        $varargs_var = $final_tuple
+
```
Yes this does. I will update the comment.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+        ):
+            # If we have reached the build_tuple
+            # we exit.
+            start_not_found = False
```
Yes it is. It is also functioning as a break for the loop, but that can be replaced with a real break.

njriasan replied:
Yes.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+    https://github.com/python/cpython/blob/a58ebcc701dd6c43630df941481475ff0f615a81/Python/compile.c#L55
+    https://github.com/python/cpython/blob/a58ebcc701dd6c43630df941481475ff0f615a81/Python/compile.c#L4442
+
+    In particular, change is imposed whenever (n_args / 2) + n_kws > 15.
```
Done.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+            _remove_assignment_definition(old_body, search_end, func_ir)
+        else:
+            # We expect to find another arg to append.
+            # The first stmt must be an add
```
Done.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+        ...
+        BUILD_CONST_KEY_MAP # Build a map
+
+    In the IR generated, the varkwarg refer
```
Done.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+        $final_tuple = $prev_tuple + $final_var
+        $varargs_var = $final_tuple
+
+    It unclear if the extra assignment at the end is always present.
```
Yes this included in the AssertionError, so I will remove this from the comments.

njriasan commented on numba/core/interpreter.py:
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
+                args = call.args
+                kws = call.kws
+                # We search for replace if a call has either vararg
+                # or varkwarg.
```
I have updated this with a longer comment explaining the meaning of this. The summary is that we investigate a call expression whenever it contains either vararg or varkwarg. varkwarg requires the bytecode changes and vararg is used for generating a nicer error message.

Please let me know if you think my more detailed comment is still unclear.

njriasan commented:
@stuartarchibald I made your requested changes. There are possibly two things that need to be looked at:

1. Should the "abundance of caution" check contain a different error like an unreachable assertion. I don't have enough confidence with the code structure to say its truly unreachable so I just added a comment.
2. Is the comment explaining when to check the call expression clear enough?

If both these changes seem fine, I think everything else should be a direct application of your feedback and we should be ready to merge. Thank you so much for the detailed review!

stuartarchibald commented on numba/core/interpreter.py:
```diff
@@ -79,6 +80,646 @@ def get_assignment_source(self, destname):
+                call = stmt.value
+                args = call.args
+                kws = call.kws
+                # We need to to check the call expression contents if
```
```suggestion
                # We need to check the call expression contents if
```

njriasan replied:
Done.

stuartarchibald commented on numba/core/interpreter.py:
```diff
@@ -79,6 +80,646 @@ def get_assignment_source(self, destname):
+                kws = call.kws
+                # We need to to check the call expression contents if
+                # it contains either vararg or varkwarg. If it contains
+                # varkwarg we need to update the actual bytecode. If it
+                # just contains vararg we don't need to update the bytecode,
```
Do you mean `bytecode` or IR?

stuartarchibald reviewed (commented):
Thanks for the fixes @njriasan, there's a typo and a query to resolve else this looks good.

stuartarchibald commented:
> @stuartarchibald I made your requested changes.

@njriasan Many thanks.

> There are possibly two things that need to be looked at:
> 
>     1. Should the "abundance of caution" check contain a different error like an unreachable assertion. I don't have enough confidence with the code structure to say its truly unreachable so I just added a comment.

I think it can be left as-is for now. Given it cannot be handled at present, if it turns out that there's use cases in practice that will trigger it, I would guess someone will make a request for it/report it. I don't think it warrants an exception containing the reportable condition text because it's already known that it might happen and it can't be handled.
 
>     2. Is the comment explaining when to check the call expression clear enough?

Couple of minor things I've noted in https://github.com/numba/numba/pull/7866#pullrequestreview-964652343 else looks good thanks.

> If both these changes seem fine, I think everything else should be a direct application of your feedback and we should be ready to merge. Thank you so much for the detailed review!

No problem, thanks for your efforts on this.

njriasan commented:
Thank you @stuartarchibald. I made the requested changes.

stuartarchibald reviewed (approved):
Thanks for the patch and fixes @njriasan, great to get this merged!
